### ISSUE TITLE
Custom Fields

### ISSUE THREAD

**Issue by xdansmith** [microsoft/bookings-samples#4]
Bookings supports defining custom fields for different services. 

Is this supported by the API currently?

**Comment by fpintos** [comment]
At this moment the API does not yet expose the custom fields.
We'll keep this issue open and then close it when we update the service with that feature.

**Comment by xdansmith** [comment]
I also noticed these don't show in Outlook either - are there plans to expose this information in Outlook in the near future?

**Comment by fpintos** [comment]
It is an interesting suggestion to post in UserVoice:
https://outlook.uservoice.com/forums/314907-microsoft-bookings

**Comment by xdansmith** [comment]
@fpintos Any update regarding adding support for Custom Fields to the API?

**Comment by fpintos** [comment]
Not yet, but it is in the plans.

**Comment by Chotimir** [comment]
@fpintos Any updates?

**Comment by fpintos** [comment]
Unfortunately, no change on this yet.
However, the raw data is available in the underlying event where booking appointment data is stored, so it is possible to get to it.
See the comments in this issue, #7, and you'll find the name of the properties; it relies on getting an /event using extended properties, but at least it unlocks the data.

**Comment by moisesynfam** [comment]
Is it possible to the get the custom fields from a service so i can add them to a custom form ?

**Comment by fpintos** [comment]
I'll forward this request to the feature team. Last time I checked the API didn't expose custom fields yet.

**Comment by willensoll** [comment]
+1

**Comment by Tronald** [comment]
Any progress on this? Those custom fields would be pretty important to be able to call. If the user is creating fields, it's because the info is needed. Would love to have this feature.

**Comment by Tronald** [comment]
Just a heads up for anyone needing to access the custom questions. I spent a long time going down this rabbit hole as it was new to me so I figured I would share my experience to save others time. The fields cannot currently be called via the bookings API, but they can be called using extended events within the Calendar API as stated by @fpintos . This isn't ideal, but it can work for many use cases.

For example:

After you have successfully called the bookings appointment, you could extract the appointment `id`. Using that `id` you could call the calendar event with extended events.

```C#
//{calendarID} = Same as your bookings calendar ID
//{eventID} = the bookings appointment id


https://graph.microsoft.com/v1.0/users/{calendarID}/events/{eventID}?$expand=singleValueExtendedProperties($filter%3Did eq 'Binary {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemAnsweredCustomQuestions')
```

The key is in the expand call. A couple things to note with this.

1. Some examples show `$filter=id eq` but this will actually fail. The `=` needs to be replaced with `%3D`. Example: `filter%3Did eq`
```
$expand=singleValueExtendedProperties($filter%3Did eq 'Binary {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemAnsweredCustomQuestions')
```

2. Call the extended properties in the api call first. I noticed if I put it after things such as date parameters it fails.

3. The GUID being passed is specified as a `Binary` meaning the the custom questions/answers will return in base64 encoded bytes. You will need to decode this to get the json value.

```C#
var base64EncodedBytes = Convert.FromBase64String(val);
string json = Encoding.UTF8.GetString(base64EncodedBytes);
```

4. Other bookings properties can be obtained in the calendar event and they are in easy to work with string formats. Here are a few.

'String {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemCustomerName' 
'String {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemCustomerEmail'
'String {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemCustomerPhone'
'String {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemCustomerNotes'

You can call multiple extended properties in the expand call like so in the API call like so.

```
$expand=singleValueExtendedProperties($filter%3Did eq 'Binary {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemAnsweredCustomQuestions' or id eq 'String {ccb88e73-951a-4b10-bb49-e6628b8d8494} Name BookingItemCustomerName')
```

[This page](https://docs.microsoft.com/en-us/graph/api/resources/extended-properties-overview?view=graph-rest-1.0) explains expanded properties a bit more.

Hope this helps anyone needing this feature.

**Comment by safari137** [comment]
Thanks @Tronald that was extremely helpful.

@fpintos is this method supported? (Pulling the data from the event extensions)  We are considering building out this functionality but are concerned about whether or not it will always work this way?

**Comment by Tronald** [comment]
No Problem @safari137 

The schema still isn't publicly documented as far I can tell (which is making finding the information I need a nightmare). Because of that fact, the GUID and Property names could certainly change so I would use this with caution. I have a hard time seeing MS change it this late in the game, but the bookings API is still in preview so I guess nothing is truly set in stone. Judging by how it appears to work though, I would venture to say that "Bookings Events" are simply an extension of a Calendar Event and stored in the Calendar as such. If that is the case the values will always be embedded in some form.

The puzzle right now seems to be figuring out how to get them all with no publicly available schema. The Graph API doesn't seem to allow you to get an event with "all extensions" either. For some reason you physically have to specify which extensions you want in the API call, even though I would rather just get them all and work backwards. If there is a way to do that, we could work in reverse to figure out the mappings (someone please let me know if there is).

@fpintos please feel free to chime in on this though as I am not privy to the inner workings of this project. Also, if we could have a schema with the GUID and Names that would be amazing (even if it is subject to change).

**Comment by coupalm** [comment]
Any idea how to extract customers info when the booking event support multiple customers (when you set maximum attendees higher than 1 in the service). It is not available in the attendees property... Booking Graph API does not expose this information also...

**Comment by rfoust** [comment]
This has been open for 3 years now...is there no way to get customers info, specifically custom fields, when there are more than 1 attendee via Graph?

**Comment by khuffman66** [comment]
I was able to extract custom field data successfully using Power Automate.

Flow Description

This flow runs nightly and parses user data from events located in a generic Outlook account calendar (this account is being used by the Bookings app to schedule appointments) then creates a CSV file stored in a Shared Documents team library.  Attached are the details of my solution.
[Bookings event details to CSV file using Power Automate.docx](https://github.com/microsoft/bookings-samples/files/5827020/Bookings.event.details.to.CSV.file.using.Power.Automate.docx)


Hope this helps!
~Kimberly

**Comment by zigagiz** [comment]
This simple issue has been open for 1259 days and it's still not resolved ...

Can we get the custom fields (questions and answers) when calling Get Appointment, please?

**Comment by khuffman66** [comment]
@zigagiz - my solution noted in [this previous post](https://github.com/microsoft/bookings-samples/issues/4#issuecomment-761900684) works; however, the TSV export in the Bookings app now includes custom fields.

**Comment by zigagiz** [comment]
@khuffman66 - I've tried Tronald's solution and it's what we'll need for our use case. I just find it baffling that custom fields aren't present in the _Get bookingAppointment_ response, but need to be retrieved via a different API and dug out from HTML...

**Comment by brad302** [comment]
@Tronald (Justin) ... thanks for the answer, very useful.  However, it's been a couple of years since you commented and provided the solution but I have to ask, how the hell did you know that those filters were the ones to use to get the relevant values?  I haven't found a scrap of documentation that provides that answer and yet here it is.

Keen to understand so if you can remember then it'd be great to share.

**Comment by Tronald** [comment]
@brad302 I am so happy my answer has been helpful. I don't remember exactly how I figured this out, but it started with reading issue #7 . Once I learned about the extended properties, I believe I basically attempted a bunch of property name guesses based on @fpintos use of 'BookingItemCustomerName' and looked at what was successfully returned. 

Thankfully, the property naming conventions were consistent so it wasn't too bad to guess my way to the answer. It would be so much nicer if Microsoft provided easy to find documentation though.

This is just one of those times where determination mixed with a bit of luck to got me to the answer.
